### PR TITLE
added ability to clear all diagnostics reproted from a IDiagnosticUpd…

### PR DESCRIPTION
**Pull request by heejaechang**
…ateSource

previously IDiagnosticUpdateSource has to clear each diagnostics it reported group by group. that was fine for IDiagnosticUpdateSource that supports incremental update, but some source such as EditAndContinue doesn't support incremental update since their errors (emit errors) come and go as a bulk (whole project). when they update, they need to update everything. so tracking things in group for incremental update is unnecessary for such source.

the new API (Source.Cleared) make it easy for source to clear all diagnostics at once it produced.

### CODE REVIEW

**Comment by heejaechang** [comment]
tagging @tmat 

you just need to raise encUpdateSource.Cleared?.Invoke(this, EventArgs.Empty) to clear all diagnostics reported from your source.

**Comment by heejaechang** [comment]
tagging @dotnet/roslyn-ide

File: src/EditorFeatures/Test/Diagnostics/DiagnosticServiceTests.cs [hunk 1]
```diff
@@ -90,6 +90,68 @@ public void TestGetDiagnostics2()
             }
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.Diagnostics)]
+        public void TestCleared()
+        {
+            using (var workspace = new TestWorkspace(TestExportProvider.ExportProviderWithCSharpAndVisualBasic))
+            {
+                var set = new ManualResetEvent(false);
```
Comment (tmat) [inline src/EditorFeatures/Test/Diagnostics/DiagnosticServiceTests.cs]: >set [](start = 20, length = 3)

`set` is an odd name for an event.

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs [hunk 2]
```diff
@@ -79,6 +79,37 @@ private void RaiseDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs args)
             }).CompletesAsyncOperation(eventToken);
         }
 
+        private void RaiseDiagnosticsCleared(object sender)
+        {
+            Contract.ThrowIfNull(sender);
+            var source = (IDiagnosticUpdateSource)sender;
```
Comment (tmat) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: I'd move this to the caller and strongly type the parameter.

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs [hunk 3]
```diff
@@ -150,12 +181,48 @@ private bool UpdateDataMap(IDiagnosticUpdateSource source, DiagnosticsUpdatedArg
             }
         }
 
+        private bool ClearDiagnosticsFromDataMapFor(IDiagnosticUpdateSource source, List<DiagnosticsUpdatedArgs> removed)
```
Comment (tmat) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: >ClearDiagnosticsFromDataMapFor [](start = 21, length = 30)

Wouldn't `ClearDiagnosticsReportedBySource` be better name?

File: src/EditorFeatures/Test/Diagnostics/DiagnosticServiceTests.cs (hunk 1 above)
Comment (heejaechang) [reply src/EditorFeatures/Test/Diagnostics/DiagnosticServiceTests.cs]: so.. wait handle then?

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs (hunk 2 above)
Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: sure.

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs (hunk 3 above)
Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/DiagnosticService.cs]: sure.

**Pull request merged**